### PR TITLE
[DEVREL-1008] Add: Voucher code generator sample /Demo-Apps.

### PR DESCRIPTION
**Pull request by teja-kummarikuntla**
## Description
A sample app for Freshdesk helps the agent to generate voucher codes for specific customers.
### TL;DR.
Support Reps will need to check in a different system to get a coupon code for a customer. Most of the inconvenience is logging into a different tool, generating a coupon code, and copy-pasting in the email, and then sending it, with this we don't do that anymore.   
#### Soft specs:
- App generates voucher code by making an SMI call to a `generateVoucher(){..}`  and renders coupon code to the app's UI.
- On a click of `Paste in Editor` button, automatically paste the coupon code at the open reply's current cursor position.
- Whenever needed app can show recent 5 coupon details generated in the UI.
- Too many vouchers stored?, on Click `Remove History` deletes the voucher code data

### CODE REVIEW

File: Demo-Apps/voucher_code_generator/app/scripts/app.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      </div>`);
+    });
+    document.querySelector("#values").innerHTML = vou.join(" ");
+
+    if (keysArr.length > 0) {
+      document.getElementById("remove_history").classList.remove("hidden");
+    }
+  }),
+    function (error) {
+      console.error(error);
```
Comment (ravirajsubramanian) [inline Demo-Apps/voucher_code_generator/app/scripts/app.js:73]: Add a constructive error message log apart from the error object itself. It might sometimes just say, "Record not found" and it will not be much helpful to troubleshoot any issues.

File: Demo-Apps/voucher_code_generator/server/lib/handle-response.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+function handleResponse(err, resp) {
```
Comment (ravirajsubramanian) [inline Demo-Apps/voucher_code_generator/server/lib/handle-response.js:1]: Remove this function and file if not used.

File: Demo-Apps/voucher_code_generator/manifest.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+      "location": {
+        "ticket_sidebar": {
+          "url": "templates/index.html",
+          "icon": "styles/images/icon.svg"
+        }
+      }
+    }
+  },
+  "dependencies": {
+    "voucher-code-generator": "1.1.1"
```
Comment (ravirajsubramanian) [inline Demo-Apps/voucher_code_generator/manifest.json:14]: This dependency looks simple and not well maintained or popular. Can we write the code in a separate util function and use it within the app instead of using this dependency?

The code can be simpler and need to be much hard-to-guess code as it is not for production.

If it's not feasible, it's fine.

File: Demo-Apps/voucher_code_generator/app/scripts/create_vocher.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+function getFieldValues() {
+  return {
+    subject: document.getElementById("voucher-subject").value,
+    description: document.getElementById("voucher-description").value,
+    discount: Number(document.getElementById("voucher-discount").value),
+    validity: document.getElementById("voucher-validity").value,
+  };
+}
+
+/* Reset from fields */
```
Comment (teja-kummarikuntla) [inline Demo-Apps/voucher_code_generator/app/scripts/create_vocher.js:15]: from -> form

File: Demo-Apps/voucher_code_generator/README.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+- Whenever needed app can show recent 5 coupon details generated in the UI.
+- Too many vouchers stored?, on Click `Remove History` deletes the voucher code data
+***
+
+## Screenshots
+
+
+<img src="./screenshots/app_face.png" hight="200px" width="250px">
+
+For other screenshots please refer to [screenshots](./screenshots) folder
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/README.md:19]: Add punctuation at EOL.

File: Demo-Apps/voucher_code_generator/app/scripts/app.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+function deleteVouchers() {
+  client.interface
+    .trigger("showConfirm", {
+      title: "Deleting Voucher Data",
+      message: "Are you sure?",
+      saveLabel: "Yes",
+      cancelLabel: "No",
+    })
+    .then(function (result) {
+      if (result.message == "Yes") {
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/scripts/app.js:87]: Use deepEqual here. `result.message === "Yes"`

File: Demo-Apps/voucher_code_generator/app/scripts/app.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        client.db.delete("vouchers").then(function () {
+          showNotify(
+            "success",
+            "Success",
+            "Vouchers History removed successfully"
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/scripts/app.js:92]: Voucher history*

File: Demo-Apps/voucher_code_generator/app/scripts/app.js [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+          );
+        }),
+          function (error) {
+            console.log(error);
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/scripts/app.js:96]: Add some description to error log. e.g., `console.log("Failed to delete vouchers", error)`

File: Demo-Apps/voucher_code_generator/app/scripts/app.js [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+      <p class="lookup-body">${dbData[element].description}</p>
+      <label class="tada-app-label text--xsmall lookup-body"><b> Discount(%) </b></label>
+      <p class="lookup-body">${dbData[element].discount}</p>
+      <label class="tada-app-label text--xsmall lookup-body"><b>Validity</b></label>
+      <p class="lookup-body">${dbData[element].validity}</p>
+      <label class="tada-app-label text--xsmall lookup-body"><b> Voucher Code </b></label>
+      <fw-label class="lookup-body" value="${dbData[element].voucher}" onClick= "pasteInEditor(\'${dbData[element].voucher}'\)" name="pasteInEditor" data-arg1="${dbData[element].voucher}" color="green"></fw-label>
+      <fw-icon name="magic-wand" size="12" color="green" onClick= "pasteInEditor(\'${dbData[element].voucher}'\)">
+      </fw-icon>
+      </div>`);
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/scripts/app.js:64]: This code is hard to read. Consider extracting the DOM element creation to a separate function that takes in `element` and returns a new DOM object or string.

Also, semantically, `<fw-input>` is probably better than using `<fw-label>` here. A label is meant to accompany an input. Your call, since changing it will break the current look and feel.

File: Demo-Apps/voucher_code_generator/app/scripts/app.js [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        }),
+          function (error) {
+            console.log(error);
+          };
+        document.getElementById("remove_history").classList.add("hidden");
+        document.getElementById("values").innerHTML = "";
+      }
+    }),
+    function (error) {
+      console.log(error);
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/scripts/app.js:103]: Add some description to error log.

File: Demo-Apps/voucher_code_generator/app/scripts/create_vocher.js [hunk 11]
```diff
@@ -0,0 +1,148 @@
+const DANGER_NOTIFICATION = "danger";
+const SUCCESS_NOTIFICATION = "success";
+const FIELDS_TO_VALIDATE = ["subject", "description", "discount", "validity"];
+
+/* Fetch data from the form fields */
+function getFieldValues() {
+  return {
+    subject: document.getElementById("voucher-subject").value,
+    description: document.getElementById("voucher-description").value,
+    discount: Number(document.getElementById("voucher-discount").value),
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/scripts/create_vocher.js:10]: Use `parseInt()` instead of directly casting to `Number`. Or, use `parseFloat()` if the returned value should be a floating point.

File: Demo-Apps/voucher_code_generator/app/scripts/create_vocher.js [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+/**
+ * Save generated voucher code in Database
+ * @param {string} subject - Voucher code subject
+ * @param {string} description - Voucher code description
+ * @param {number} discount - Discount value in percentage
+ * @param {string} validity - Voucher code time period in data scale
+ * @param {string} voucher - generated voucher code
+ */
+function saveVoucherToDb(subject, description, discount, validity, voucher) {
+  let createObject = new Object();
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/scripts/create_vocher.js:38]: `let createObject = {};`

File: Demo-Apps/voucher_code_generator/app/scripts/create_vocher.js [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+  createObject[`voucher_${id}`] = {
+    subject,
+    description,
+    discount,
+    validity,
+    voucher,
+  };
+  client.db.update("vouchers", "set", createObject),
+    function (error) {
+      console.error(error);
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/scripts/create_vocher.js:49]: Add some description to error log.

File: Demo-Apps/voucher_code_generator/app/scripts/create_vocher.js [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+  document
+    .getElementById("create-voucher")
+    .addEventListener("click", function () {
+      if (validateFields()) {
+        let vSubject = document.getElementById("voucher-subject").value;
+        let vDescription = document.getElementById("voucher-description").value;
+        let vDiscount = Number(
+          document.getElementById("voucher-discount").value
+        );
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/scripts/create_vocher.js:76]: Use `parseInt()` or `parseFloat()` instead.

File: Demo-Apps/voucher_code_generator/app/scripts/create_vocher.js [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    voucher,
+  };
+  client.db.update("vouchers", "set", createObject),
+    function (error) {
+      console.error(error);
+    };
+}
+
+/* Element event listeners */
+function addListeners() {
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/scripts/create_vocher.js:54]: There's a _lot_ happening in this function. Extract logical chunks into separate, well-named functions.

File: Demo-Apps/voucher_code_generator/app/scripts/util.js [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+    title: title,
+    message: message,
+  }),
+    function (error) {
+      console.error("Error from showNotify:", title, error);
+    };
+}
+/* Return Unique Identifier */
+function generateUuid() {
+  return Math.floor(Math.random() * 1000000000);
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/scripts/util.js:14]: This is not generate a fixed length number on each call. It will not guarantee a UUID at least. Good enough for demo purposes though.

Tip: Add an inline code comment saying something like: `// Replace this with a proper UUID generator for production apps`.

Tip: Maybe don't call the function `generateUuid()`, but call it `generateId()` instead?

File: Demo-Apps/voucher_code_generator/app/scripts/util.js [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+/**
+ * Perform copy data to clipboard
+ * @param {string} text - data needed to copy.
+ */
+function copyToClipboard(text) {
+  var input = document.body.appendChild(document.createElement("input"));
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/scripts/util.js:22]: Be const about `const`. Replace `var` here :-)

File: Demo-Apps/voucher_code_generator/app/styles/style.css [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+.alignment-middle {
+  text-align: center;
+}
+
+.button-middle,
+.align-middle {
+  width: 40%;
+  margin-left: 30%;
+  margin-right: 30%;
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/styles/style.css:40]: This is a good case for using `margin-left: auto; margin-right: auto;`. Your call.

File: Demo-Apps/voucher_code_generator/app/styles/style.css [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+
+.action-icon {
+  cursor: pointer;
+  margin-left: 10px;
+  margin-top: 25px;
+}
+
+.create-vocher-logo {
+  content: url("../styles/images/vocher-icon.png");
+  height: 23%;
```
Comment (kaustavdm) [inline Demo-Apps/voucher_code_generator/app/styles/style.css:84]: Using percentile `height` can backfire badly. Use fixed length in `px` or `rem` if possible.

File: Demo-Apps/voucher_code_generator/manifest.json (hunk 3 above)
Comment (kaustavdm) [reply Demo-Apps/voucher_code_generator/manifest.json:14]: I don't think this dependency is used at all. Given there is the `generateUuid()` function which seems to be the one generating voucher code random numbers currently,

**Review by kaustavdm — CHANGES_REQUESTED** [review]
Good job with the PR, @teja-kummarikuntla. I have flagged a few tweaks here and there. Overall:

- Use `Markdownlint` in your editor if possible to help keep an even formatting of Markdown files
- Add a helpful string describing every `console.log()` and `console.error()`. See if any error scenario should be notified explicitly, in which case use notifications for those scenarios.
- Don't let functions grow too big
- Add inline comments wherever you are performing some action crucial to this demo, or where the code looks complex enough

File: Demo-Apps/voucher_code_generator/app/scripts/util.js [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+    .trigger("setValue", { id: "editor", text: value })
+    .then(function () {
+      showNotify(SUCCESS_NOTIFICATION, "Success:", "Pasted in Editor");
+    }),
+    function (error) {
+      console.error(error);
+    };
+}
+/**
+ * Perform filed validation on create_vocher.html
```
Comment (teja-kummarikuntla) [inline Demo-Apps/voucher_code_generator/app/scripts/util.js:58]: filed -> field

File: Demo-Apps/voucher_code_generator/app/scripts/app.js (hunk 9 above)
Comment (teja-kummarikuntla) [reply Demo-Apps/voucher_code_generator/app/scripts/app.js:64]: > Also, semantically, <fw-input> is probably better than using <fw-label> here. A label is meant to accompany an input. Your call, since changing it will break the current look and feel.

Just followed with existing apps and how they used `label`, it looks fair as of now, and the card is breaking while I try chaining shifting to `input` 🤭

**Review by ravirajsubramanian — APPROVED** [review]

**Review by kaustavdm — APPROVED** [review]
